rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md

```diff
@@ -0,0 +1,182 @@
+# RFC 6330 - 2021-04-07 - Emitting multiple log events from remap transform
+
+This RFC describes an approach for emitting multiple log events from a remap transform.
+
+[Credit to @JeanMertz](https://github.com/timberio/vector/pull/7038#discussion_r609476064) for this revised approach.
+
+## Scope
+
+In:
+
+- Turning a log event into multiple log events
+
+Out:
+
+- Turning a log event into multiple metrics
+- Turning a metric into multiple metrics
+- Turning a metric into multiple logs
+
+## Motivation
+
+Currently the remap transform can only modify an event as it is passing through. Users have asked [1][] [2][] to be able to transform an incoming event into multiple events. This is useful, for example, when parsing an incoming JSON payload that contains an array for which you'd like to publish an event for each element.
+
+Currently, users are restricted to using the Lua or WASM transform for this, [introducing a substantial bottleneck](https://user-images.githubusercontent.com/316880/105531520-d3643200-5cbf-11eb-8b35-fe1c99e5c254.png).
+
+## Internal Proposal
+
+The proposal is to add an `explode` transform that makes use of remap to emit a set of events from one input event by requiring the remap script to resolve to an array. For each element of the array, a separate event will be published.
+
+Example input:
+
+```json
+{ "events": [{ "message": "foo" }, { "message": "bar" }] }
+```
+
+Transform config:
+
+```toml
+[transforms.explode]
+type = "explode"
+source = "array!(.events) ?? []" # will be typechecked at compile-time
+```
+
+Ouput:
+
+```json
+{"message": "foo"}
+{"message": "bar"}
+```
+
+Support for iteration as part of [#6031](https://github.com/timberio/vector/issues/6031) will allow for users to do things like map fields onto each element. An example might look something like:
+
+Input:
+
+```json
+{ "host": "foobar", "events": [{ "message": "foo" }, { "message": "bar" }] }
+```
+
+Transform config (actual mapping syntax TBA):
+
+```toml
+[transforms.explode]
+type = "explode"
+source = "map(array!(.events), |event| event.host = .host) ?? []"
+```
+
+Output:
+
+```json
+{"host": "foobar", "message": "foo"}
+{"host": "foobar", "message": "bar"}
+```
```

Comment:
Is this relevant to mention in this RFC? It seems orthogonal.

Also, it might actually be _more performant_ to do this in a separate `remap` transform, which can then/eventually run in parallel for multiple events.

Reply:
True, but how would they get the fields from the original source event in the later remap transform? I think they'd need to do the mapping in the `explode` transform.